In [1]:
import pandas as pd
import numpy as np
import requests, copy
from pprint import pprint
from bs4 import BeautifulSoup

### Create Initial DataFrame from scraping http://www.prosportstransactions.com/

In [7]:
df_json = pd.read_json("../data/pst_nba_injuries_2020_2021.txt")
df = df_json.copy()
df.head(5)

Date   Team        Healed             Injured  \
0 2020-03-11  76ers   Joel Embiid                       
1 2020-03-11  Bucks  Eric Bledsoe                       
2 2020-03-11  Bucks   George Hill                       
3 2020-03-11  Bulls                Chandler Hutchison   
4 2020-03-11   Heat                      Jimmy Butler   

                                        Notes  
0                           activated from IL  
1                           activated from IL  
2                           activated from IL  
3  surgery on right shoulder (out for season)  
4       placed on IL with left big toe injury

### Reformat the DataFrame

In [8]:

df['Player'] = df['Healed'] + df['Injured']
df['Status'] = np.where(df['Healed'] == '', 'Injured', 'Healed')
df = df[['Date', 'Team', 'Player', 'Status', 'Notes']]
df.head(10)

Date     Team              Player   Status  \
0 2020-03-11    76ers         Joel Embiid   Healed   
1 2020-03-11    Bucks        Eric Bledsoe   Healed   
2 2020-03-11    Bucks         George Hill   Healed   
3 2020-03-11    Bulls  Chandler Hutchison  Injured   
4 2020-03-11     Heat        Jimmy Butler  Injured   
5 2020-03-11     Heat         Jae Crowder   Healed   
6 2020-03-11     Heat         Tyler Herro   Healed   
7 2020-03-11  Hornets        Terry Rozier  Injured   
8 2020-03-11     Jazz     Emmanuel Mudiay  Injured   
9 2020-03-11     Jazz         Rudy Gobert  Injured   

                                        Notes  
0                           activated from IL  
1                           activated from IL  
2                           activated from IL  
3  surgery on right shoulder (out for season)  
4       placed on IL with left big toe injury  
5                          returned to lineup  
6                           activated from IL  
7                   placed on IL with illness  
8                               illness (DTD)  
9     COVID-19 coronavirus (out indefinitely)

### Extract players out for COVID or "Health and Safety Protocols"

In [9]:
df_cv19 = df[np.logical_or(df['Notes'].str.contains('COVID'), df['Notes'].str.contains('protocol'))]
df_cv19

Date     Team            Player   Status  \
9    2020-03-11     Jazz       Rudy Gobert  Injured   
19   2020-03-12     Jazz  Donovan Mitchell  Injured   
21   2020-03-14  Pistons    Christian Wood  Injured   
22   2020-03-17     Nets      Kevin Durant  Injured   
24   2020-03-19  Celtics      Marcus Smart  Injured   
...         ...      ...               ...      ...   
2846 2021-05-09  Pistons    Hamidou Diallo  Injured   
2957 2021-05-14  Thunder         Josh Hall  Injured   
2969 2021-05-15     Heat    Omer Yurtseven  Injured   
3031 2021-05-16    Spurs   DaQuan Jeffries  Injured   
3041 2021-05-18   Pacers      Caris LeVert  Injured   

                                                  Notes  
9               COVID-19 coronavirus (out indefinitely)  
19                           COVID-19 coronavirus (DTD)  
21                           COVID-19 coronavirus (DTD)  
22                NBA health and safety protocols (DTD)  
24              COVID-19 coronavirus (out indefinitely)  
...                                                 ...  
2846  placed on IL with NBA health and safety protocols  
2957  placed on IL with NBA health and safety protoc...  
2969  placed on IL with NBA health and safety protocols  
3031  placed on IL with NBA health and safety protoc...  
3041  placed on IL with NBA health and safety protoc...  

[240 rows x 5 columns]

240 instances of players being held out due to COVID protocols since 03/11/2020 (The day Rudy Gobert tested positive and the league shut down)

### Just COVID data for 2020-2021 NBA Season

In [10]:
df_cv19_2021 = df_cv19[df_cv19['Date'] >= '12-01-2020']
df_cv19_2021

Date     Team                           Player   Status  \
415  2020-12-22  Blazers                    Nassir Little  Injured   
423  2020-12-22    Hawks                      Rajon Rondo  Injured   
468  2020-12-23  Rockets  Kenyon Martin Jr. / K.J. Martin  Injured   
472  2020-12-25    Bucks                   Mamadi Diakite  Injured   
475  2020-12-25  Rockets                     Ben McLemore  Injured   
...         ...      ...                              ...      ...   
2846 2021-05-09  Pistons                   Hamidou Diallo  Injured   
2957 2021-05-14  Thunder                        Josh Hall  Injured   
2969 2021-05-15     Heat                   Omer Yurtseven  Injured   
3031 2021-05-16    Spurs                  DaQuan Jeffries  Injured   
3041 2021-05-18   Pacers                     Caris LeVert  Injured   

                                                  Notes  
415                                      COVID-19 (DTD)  
423   placed on IL with NBA health and safety protocols  
468   placed on IL with NBA health and safety protocols  
472   placed on IL with NBA health and safety protoc...  
475                                  COVID-19 protocols  
...                                                 ...  
2846  placed on IL with NBA health and safety protocols  
2957  placed on IL with NBA health and safety protoc...  
2969  placed on IL with NBA health and safety protocols  
3031  placed on IL with NBA health and safety protoc...  
3041  placed on IL with NBA health and safety protoc...  

[230 rows x 5 columns]

230 instances of players being held out due to COVID protocols during 2020-2021 NBA season

### Unique Players

In [15]:
players = df_cv19_2021['Player'].unique()
len(players)

182

182 Unique Players were held out due to COVID protocols during 2020-2021 NBA season

### Generate basketball-reference.com IDs

In [18]:
bbref_id_list = []
for player in players:
    player_copy = player
    # cleaning name form
    if '.' in player:
        player_copy = player_copy.replace('.', '')
    if 'ö' in player:
        player_copy = player_copy.replace('ö', 'o')
    if 'Bam Adebayo' in player_copy:
        player_copy = 'Bam Adebayo'
    
    player_split = player_copy.lower().split()
    
    # Typical ID format
    bbref_id = player_split[1][:5] + player_split[0][:2] + '01'
    # Special Cases that threw errors in test
    if player_split[1] == 'kleber':
        bbref_id = 'klebima01'
    if player_split[1] == 'ntilikina':
        bbref_id = 'ntilila01'
    bbref_id_list.append([player, bbref_id])

    ### Test to check if ID's are correct
    # url = f'https://www.basketball-reference.com/players/{bbref_id[0]}/{bbref_id}.html'
    # r = requests.get(url)
    # if r.status_code == 200:
    #     print(player, bbref_id, r.status_code)
    # else:
    #     print(player.upper())

In [19]:
df_bbref_id = pd.DataFrame(bbref_id_list, columns=['Player', 'BBRef_ID'])

### Merging BBRef IDs with DataFrame

In [20]:
df_main = pd.merge(left=df_cv19_2021, right=df_bbref_id, how='left', left_on=df_cv19_2021.Player, right_on=df_bbref_id.Player)
df_main

key_0       Date     Team  \
0                      Nassir Little 2020-12-22  Blazers   
1                        Rajon Rondo 2020-12-22    Hawks   
2    Kenyon Martin Jr. / K.J. Martin 2020-12-23  Rockets   
3                     Mamadi Diakite 2020-12-25    Bucks   
4                       Ben McLemore 2020-12-25  Rockets   
..                               ...        ...      ...   
225                   Hamidou Diallo 2021-05-09  Pistons   
226                        Josh Hall 2021-05-14  Thunder   
227                   Omer Yurtseven 2021-05-15     Heat   
228                  DaQuan Jeffries 2021-05-16    Spurs   
229                     Caris LeVert 2021-05-18   Pacers   

                            Player_x   Status  \
0                      Nassir Little  Injured   
1                        Rajon Rondo  Injured   
2    Kenyon Martin Jr. / K.J. Martin  Injured   
3                     Mamadi Diakite  Injured   
4                       Ben McLemore  Injured   
..                               ...      ...   
225                   Hamidou Diallo  Injured   
226                        Josh Hall  Injured   
227                   Omer Yurtseven  Injured   
228                  DaQuan Jeffries  Injured   
229                     Caris LeVert  Injured   

                                                 Notes  \
0                                       COVID-19 (DTD)   
1    placed on IL with NBA health and safety protocols   
2    placed on IL with NBA health and safety protocols   
3    placed on IL with NBA health and safety protoc...   
4                                   COVID-19 protocols   
..                                                 ...   
225  placed on IL with NBA health and safety protocols   
226  placed on IL with NBA health and safety protoc...   
227  placed on IL with NBA health and safety protocols   
228  placed on IL with NBA health and safety protoc...   
229  placed on IL with NBA health and safety protoc...   

                            Player_y   BBRef_ID  
0                      Nassir Little  littlna01  
1                        Rajon Rondo  rondora01  
2    Kenyon Martin Jr. / K.J. Martin  martike01  
3                     Mamadi Diakite  diakima01  
4                       Ben McLemore  mclembe01  
..                               ...        ...  
225                   Hamidou Diallo  diallha01  
226                        Josh Hall   halljo01  
227                   Omer Yurtseven  yurtsom01  
228                  DaQuan Jeffries  jeffrda01  
229                     Caris LeVert  leverca01  

[230 rows x 8 columns]